packet-solutions/xdp_prog_kern.c

```diff
@@ -0,0 +1,419 @@
+ * if it exists, otherwise push a new one with ID 1
+ */
+SEC("xdp_vlan_swap")
+int xdp_vlan_swap_func(struct xdp_md *ctx)
+{
+	void *data_end = (void *)(long)ctx->data_end;
+	void *data = (void *)(long)ctx->data;
+
+        /* These keep track of the next header type and iterator pointer */
+	struct hdr_cursor nh;
+	int nh_type;
+        nh.pos = data;
+
+	struct ethhdr *eth;
+	nh_type = parse_ethhdr(&nh, data_end, &eth);
+        if (nh_type < 0)
+                return XDP_PASS;
+
+        if (proto_is_vlan(eth->h_proto))
+                vlan_tag_pop(ctx, eth);
+        else
+                vlan_tag_push(ctx, eth, 1);
+
+        return XDP_PASS;
+}
+
+static __always_inline void swap_src_dst_mac(struct ethhdr *eth)
+{
+	__u8 h_tmp[ETH_ALEN];
+	memcpy(h_tmp, eth->h_source, ETH_ALEN);
+	memcpy(eth->h_source, eth->h_dest, ETH_ALEN);
+	memcpy(eth->h_dest, h_tmp, ETH_ALEN);
+}
+
+static __always_inline void swap_src_dst_ipv6(struct ipv6hdr *ipv6)
+{
+	struct in6_addr tmp = ipv6->saddr;
+	ipv6->saddr = ipv6->daddr;
+	ipv6->daddr = tmp;
+}
+
+static __always_inline void swap_src_dst_ipv4(struct iphdr *iphdr)
+{
+	__be32 tmp = iphdr->saddr;
+	iphdr->saddr = iphdr->daddr;
+	iphdr->daddr = tmp;
+}
+
+static __always_inline __u16 csum16_add(__u16 csum, __u16 addend)
+{
+	csum += addend;
+	return csum + (csum < addend);
+}
+
+/* Solution to packet03/assignment-1 */
+SEC("xdp_icmp_echo")
+int xdp_icmp_echo_func(struct xdp_md *ctx)
+{
+	void *data_end = (void *)(long)ctx->data_end;
+	void *data = (void *)(long)ctx->data;
+	struct hdr_cursor nh;
+	struct ethhdr *eth;
+	int eth_type;
+	int ip_type;
+	int icmp_type;
+	struct iphdr *iphdr;
+	struct ipv6hdr *ipv6hdr;
+	__u16 echo_reply, m0, m1;
+	struct icmphdr *icmphdr;
+	__u32 action = XDP_PASS;
+
+	/* These keep track of the next header type and iterator pointer */
+	nh.pos = data;
+
+	/* Parse Ethernet and IP/IPv6 headers */
+	eth_type = parse_ethhdr(&nh, data_end, &eth);
+	if (eth_type == ETH_P_IP) {
+		ip_type = parse_iphdr(&nh, data_end, &iphdr);
+		if (ip_type != IPPROTO_ICMP)
+			goto out;
+	} else if (eth_type == ETH_P_IPV6) {
+		ip_type = parse_ip6hdr(&nh, data_end, &ipv6hdr);
+		if (ip_type != IPPROTO_ICMPV6)
+			goto out;
+	} else {
+		goto out;
+	}
+
+	icmp_type = parse_icmphdr(&nh, data_end, &icmphdr);
+	if (eth_type == ETH_P_IP && icmp_type == ICMP_ECHO) {
+		/* Swap IP source and destination */
+		swap_src_dst_ipv4(iphdr);
+		echo_reply = ICMP_ECHOREPLY;
+	} else if (eth_type == ETH_P_IPV6 && icmp_type == ICMPV6_ECHO_REQUEST) {
+		/* Swap IPv6 source and destination */
+		swap_src_dst_ipv6(ipv6hdr);
+		echo_reply = ICMPV6_ECHO_REPLY;
+	} else {
+		goto out;
+	}
+
+	/* Swap Ethernet source and destination */
+	swap_src_dst_mac(eth);
+
+	/* Patch the packet and update the checksum */
+	m0 = * (__u16 *) icmphdr;
+	icmphdr->type = echo_reply;
+	m1 = * (__u16 *) icmphdr;
+	icmphdr->checksum = ~(csum16_add(csum16_add(~icmphdr->checksum, ~m0), m1));
```

Comment:
Sure, I will replace this with `bpf_csum_diff`. I wanted to use a hand-made implementation so that we can precisely see that the new sum is just `~(~sum +  ~m0 +  m1)`